src/presentation/http/router/upload.ts

```diff
@@ -0,0 +1,113 @@
+import type { FileTypes } from '@domain/entities/file';
+import type { NotePublicId } from '@domain/entities/note.js';
+import type FileUploaderService from '@domain/service/fileUploader.service.js';
+import type { MultipartFile, MultipartValue } from '@fastify/multipart';
+import type { FastifyPluginCallback } from 'fastify';
+import useNoteResolver from '../middlewares/note/useNoteResolver.js';
+import type NoteService from '@domain/service/note.js';
+
+/**
+ * Interface for upload router options
+ */
+interface UploadRouterOptions {
+  /**
+   * File uploader service
+   */
+  fileUploaderService: FileUploaderService;
+
+  /**
+   * Note service instance
+   */
+  noteService: NoteService;
+}
+
+/**
+ * Interface for upload file body
+ */
+interface UploadFileBody {
+  /**
+   * File to upload
+   */
+  file: MultipartFile;
+
+  /**
+   * File type
+   */
+  type: MultipartValue<FileTypes>;
+
+  /**
+   * Note public id, if file is related to a note
+   */
+  notePublicId?: MultipartValue<NotePublicId>;
+}
+
+const UploadRouter: FastifyPluginCallback<UploadRouterOptions> = (fastify, opts, done) => {
+  const { fileUploaderService } = opts;
+  const { noteService } = opts;
+
+  /**
+   * Prepare note id resolver middleware
+   * It should be used in routes that accepts note public id
+   */
+  const { softNoteResolverForUploadFileRequest } = useNoteResolver(noteService);
+
+  fastify.post<{
+    Body: UploadFileBody;
+  }>('/', {
+    config: {
+      policy: [
+        'authRequired',
+        'userCanUpload',
+      ],
+    },
+    preHandler: [
+      softNoteResolverForUploadFileRequest,
+    ],
+  }, async (request, reply) => {
+    const { userId } = request;
+    const { note } = request;
+
+    const uploadedFileKey = await fileUploaderService.uploadFile({
+      data: await request.body.file.toBuffer(),
+      type: request.body.type.value,
+      mimetype: request.body.file.mimetype,
+      name: request.body.file.filename,
+    }, {
+      userId: userId as number,
+      noteId: note?.id as number,
```

Comment:
note can't be undefined here

Reply:
it can be undefined in case if file is not part of note, for example avatar image

Reply:
I'd suggest organizing it via the separate argument:

```js
uploadFile(type, file, location, meta)
```

where
- `type` — file type
- `file` — file info `{buffer, mime, name}`
- `location` — file location that depends on type. For example `{ noteId }` for note-related types
- `meta` — `{userId}`

Reply:
changed